components/prism-factor.js

```diff
@@ -0,0 +1,277 @@
+					pattern: /(\s)[^:\r\n\t ]+/,
+					lookbehind: true
+				}
+			}
+		},
+
+		/* this description of stack effect literal syntax is not complete and not as specific as theoretically possible
+		 	trying to do better is more work and regex-computation-time than it's worth though.
+			- we'd like to have the "delimiter" parts of the stack effect [ (, --, and ) ] be a different (less-important or comment-like) colour to the stack effect contents
+			- we'd like if nested stack effects were treated as such rather than just appearing flat
+			- we'd like if the following variable name conventions were recognised specifically:
+				special row variables = ..a b..
+				nested quotation variables end with a colon = quot: ( a: ( -- ) -- b )
+				word throws unconditional error = *
+				any other word-like variable name = a ? q' etc
+
+			https://docs.factorcode.org/content/article-effects.html
+
+			these are pretty complicated to highlight properly without a real parser, and therefore out of scope
+		'stack-effect-old': {
+			pattern: /(^|\s)(?:call|execute|eval)?\((?:\s+[^"\r\n\t ]\S*)*?\s+--(?:\s+[^"\n\t ]\S*)*?\s+\)(?=\s|$)/,
+			lookbehind: true,
+			alias: 'operator',
+		},
+
+		TODO:
+		'stack-effect-variable': {
+			{ pattern: / /, lookbehind: true, alias: 'operator' },
+		},
+		*/
+
+		// current solution is not great
+		'stack-effect-open': {
+			pattern: /(^|\s)(?:call|execute|eval)?\((?=\s)/,
+			lookbehind: true,
+			alias: 'operator' },
+		'stack-effect-middle': {
+			pattern: /(\s)--(?=\s)/,
+			lookbehind: true,
+			alias: 'operator' },
+		'stack-effect-close': {
+			pattern: /(\s)\)(?=\s|$)/,
+			lookbehind: true,
+			alias: 'operator' },
+
+		combinators: [],
+
+		'kernel-builtin': {
+			pattern: null,
+			lookbehind: true,
+			alias: 'variable'
+		},
+
+		'sequences-builtin': {
+			pattern: null,
+			lookbehind: true,
+			alias: 'variable'
+		},
+
+		'math-builtin': {
+			pattern: null,
+			lookbehind: true,
+			alias: 'variable'
+		},
+
+		'constructor-word': {
+			// <array> but not <=>
+			pattern: /(^|\s)<(?!=+>|-+>)[^\r\n\t ]+>(?=\s|$)/,
+			lookbehind: true,
+			alias: 'keyword'
+		},
+
+		'other-builtin-syntax': {
+			pattern: null,
+			lookbehind: true,
+			alias: 'operator'
+		},
+
+		'conventionally-named-word': {
+			// for certain word naming conventions like +lt+, sequence?, readers>>, >>setters, writers<<
+			pattern: /(^|\s)(?:set-[^\r\n\t ]+|change-[^\r\n\t ]+|>[^>\r\n\t ]+|[^:>\r\n\t ]+>|[^>\r\n\t ]+>[^>\r\n\t ]+|\+[^\+\r\n\t ]+\+|[^?\r\n\t ]+\?|[^>\r\n\t ]+>>|>>[^>\r\n\t ]+|[^<\r\n\t ]+<<|\?[^?\r\n\t ]+|\([^()\r\n\t ]+\)|[^!\r\n\t ]+!)(?=\s|$)/,
+			lookbehind: true,
+			alias: 'keyword'
+		},
+
+		'colon-syntax': {
+			pattern: /(^|\s)(?:[A-Z]*#?):{1,2}\s+(?:(?!;)|;\S+|[^\r\n\t; ]\S*)(?=\s|$)/,
+			lookbehind: true,
+			greedy: true,
+			alias: 'function'
+		},
+
+		'semicolon-or-setlocal': {
+			pattern: /(\s)(?:;|:>)(?=\s|$)/,
+			lookbehind: true,
+			alias: 'function'
+		},
+
+		'curly-brace-literal-open': {
```

Comment:
Readability, simplicity and naming them differently were all I had in mind. I'll combine them however.

Reply:
The `lookbehind: true` option behaves in an unexpected manner with `(?: (lookbehind 1)pattern | (lookbehind 2)pattern )` (i.e. `lookbehind 2` is not treated as lookbehind by Prism; apparently `lookbehind: true` effectively drops the first capturing group from the entire pattern, rather than dropping the first captured group in the match). 

Whatever the case, it's not possible to combine these patterns into a single pattern and still use `lookbehind: true` for each pattern's preceding whitespace, but I have combined them into arrays.